### PR TITLE
Use fragment instance parallel for bucket shuffle and colocate join

### PR DESCRIPTION
**Pull request by ZiheLiu**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Problem Summary：

Local bucket shuffle and colocate join should use fragment instance parallel instead of pipeline parallel, to avoid local shuffle and too large in-filter in the left scan node, even if local bucket shuffle join is succeeded by a **broadcast join**.
Otherwise, the merged in-filter may be too large to push down to the storage level.

## Test
After this optimization, query21 of TPC-DS 100g on 4 BE each of which contains 16 vCPU decrease from 400ms to 160ms.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 12 / 18 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|18|66.67%|

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 12 / 18 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|12|18|66.67%|

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/JoinTest.java [hunk 1]
```diff
@@ -2158,4 +2163,97 @@ public void testColocateJoinWithDiffPredicateOrders() throws Exception {
...
+            @Mock
+            public int getAvgNumOfHardwareCoresOfBe() {
+                return numCPUs;
+            }
+        };
+
+        // Enable DopAutoEstimate.
+        SessionVariable sessionVariable = connectContext.getSessionVariable();
+        Assert.assertNotNull(sessionVariable);
+        sessionVariable.setEnablePipelineEngine(true);
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/JoinTest.java:2180]: The session variable value need to set origin value.

Comment (ZiheLiu) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/JoinTest.java:2180]: Yes, it is set to origin value at the tail of this test function.

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 2]
```diff
@@ -1752,7 +1764,11 @@ public PlanFragment visitPhysicalHashJoin(OptExpression optExpr, ExecPlan contex
                     context.getFragments().add(leftFragment);
 
                     leftFragment.mergeQueryGlobalDicts(rightFragment.getQueryGlobalDicts());
-                    estimateDopOfBroadcastJoinInPipeline(leftFragment);
+                    if (distributionMode.equals(HashJoinNode.DistributionMode.COLOCATE)) {
```
Comment (satanson) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1767]: can this condition expr exclude fragment instance contains HashJoinNode(shuffle)->HashJoinNode(bucket shuffle) ?

Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1767]: @satanson No, `HashJoinNode(shuffle)->HashJoinNode(bucket shuffle)` will also set dop=1 and num_fragment_instance=CPUs/2.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|16|18|88.89%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Review by murphyatwork — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 18 / 20 (90.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|16|18|88.89%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Review by kangkaisen — DISMISSED** [review]

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Review by satanson — APPROVED** [review]

**Comment by ZiheLiu** [comment]
@Mergifyio backport branch-2.2

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 16 / 18 (88.89%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|

**Pull request merged**